Compute correlations -- **H2**
```{r}
# compute correlations between h2/ch4 ratio and species abundances
ch4_corr <- sapply(hydro_meth_species_tx[,9:ncol(hydro_meth_species_tx)], function(species) {
  test_result <- cor.test(hydro_meth_species_tx$CH4, species, method = 'spearman') # method = 'pearson'
  c(Estimate = test_result$estimate, Pvalue = test_result$p.value)
}, simplify = "data.frame")

# Convert results list to dataframe
ch4_corr_results <- as.data.frame(ch4_corr)
ch4_corr_results <- data.frame(t(ch4_corr_results))
colnames(ch4_corr_results) <- c("Correlation", "Pvalue")

# Add species names (rownames) as a new column
ch4_corr_results <- rownames_to_column(ch4_corr_results, var = "Species")
# rownames(ch4_corr_results) <- NULL
head(ch4_corr_results)

```
Save abundance correlations as .csv file
```{r}
write.csv(ch4_corr_results, "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/ch4_corr_results.csv", row.names = FALSE)

```
Rename taxa to simply Species name
```{r}
# Define a function to find the most specific taxon
find_most_specific_taxon <- function(taxon_string) {
  # Split the string into individual taxon
  split_taxon <- strsplit(taxon_string, "\\.")[[1]]

  # Remove the level identifiers
  named_taxon <- gsub("^.*__ ", "", split_taxon)

  # Reverse the order
  rev_taxon <- rev(named_taxon)

  # Find the first taxon that doesn't say 'uncultured' or 'metagenome'
  specific_taxon <- rev_taxon[!rev_taxon %in% c('uncultured',
                                                'unidentified',
                                                'uncultured_bacterium',
                                                'metagenome',
                                                'uncultured_organism',
                                                'human_gut',
                                                'gut_metagenome',
                                                '_bacterium')][1]

  return(specific_taxon)
}

species_c4m <- ch4_corr_results
species_c4m$Species <- ch4_corr_results$Species

# View the updated data frame
head(species_c4m)

# Remove extra "__." 
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)
species_c4m$Species <- gsub("\\.__$", "", species_c4m$Species)

# Remove "." where it's not followed by any more characters
species_c4m$Species <- gsub("\\.$", "", species_c4m$Species)

# Remove "." where followed by a single "_"
species_c4m$Species <- gsub("\\._(?!_)", "_", species_c4m$Species, perl=TRUE)

# Remove "." when preceded by two "__"
species_c4m$Species <- gsub("(?<=__)\\.", "", species_c4m$Species, perl=TRUE)

# Add a space after each of the specified patterns
species_c4m$Species <- gsub("(d__|p__|c__|o__|f__|g__|s__)", "\\1 ", species_c4m$Species)

# Remove "." from species/genera like UCG.003 so it reads UCG003
species_c4m$Species <- gsub("([a-zA-Z])\\.([0-9])", "\\1\\2", species_c4m$Species)

# Check
head(species_c4m$Species)

# Apply function to Species count table
species_c4m$Species <- sapply(species_c4m$Species, find_most_specific_taxon)

# Clean up special characters left in names
species_c4m$Species <- gsub("\\[|\\]", "", species_c4m$Species)

# Replace underscores with spaces
species_c4m$Species <- gsub("_", " ", species_c4m$Species)

# Replace "." in sp.
species_c4m$Species <- gsub("sp$", "sp.", species_c4m$Species)

ch4_corr_results <- species_c4m

```
Correlation log plot -- **H2_toCH4**
```{r}
library(ggrepel)

ggplot(ch4_corr_results, aes(x=Correlation, y=-log10(Pvalue))) + geom_point() + labs(x="Correlation", y="-log10(Pvalue)") + geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red")

# Create a new column in the dataframe to indicate significance
ch4_corr_results$Significant <- ifelse(ch4_corr_results$Pvalue < 0.05, "Significant", "Not significant")

# Create the volcano plot
p_ch4_corr <- ggplot(ch4_corr_results, aes(x=Correlation, y=-log10(Pvalue), color=Significant)) + 
  geom_point() + 
  geom_hline(yintercept=-log10(0.05), linetype="dashed", color = "red") +
  labs(x="Spearman Ï", y="-log10(Pvalue)", color="Significance") + 
  theme_bw() + 
  geom_text_repel(data = subset(ch4_corr_results, Pvalue < 0.05), aes(label=Species), 
                  size=3, box.padding = unit(0.35, "lines"), max.overlaps = 13, color = "black") +
  ggtitle("") + # "Methane"
  theme(legend.title = element_blank(), # legend.title = element_text(size = 12) # changes the legend title font size
        legend.text = element_text(size = 12))
p_ch4_corr

# # Correct 'Species' names if needed
# ch4_corr_results[317, "Species"] <- "Defluviitaleaceae_UCG.011"
# ch4_corr_results[227, "Species"] <- "Ruminococcaceae"
# ch4_corr_results[100, "Species"] <- "Enterobacteriaceae"
# ch4_corr_results[83, "Species"] <- "Erysipelotrichaceae UCG 003"

```

```{r}
# Save plots as .png images
ggsave(filename = "~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses/results/h2_ch4_corr/p_ch4_abund_corr.png", plot = p_ch4_corr, width = 10, height = 6, dpi = 300)

```