---
title: "Microbiome-OC"
author: "Fernanda Terrazas"
date: "`r Sys.Date()`"
output: pdf_document
---
```{r Setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```
Load Packages
```{r Load Packages}
my_packages <- c("FSA", "ggplot2","tidyverse", "corrplot", "ggpubr", "psych","vegan", "nlme", "metap", "caTools", "randomForest", "ROCR", "phyloseq", "car", "ape", "tidyr","gridExtra", "lme4")
lapply(my_packages, library, character.only = TRUE)

```
Check for RStudio and packages versions
```{r Load Packages}
# RStudio
library(rstudioapi)
rstudioapi::versionInfo()$version

# R packages
packageVersion("coda4microbiome")
packageVersion("ggplot2")
packageVersion("vegan")
packageVersion("lme4")

```
Set working directory
```{r Load Packages}
setwd("~/Documents/Programming/Microbiome/microbiome-ocp/data-analyses")

```
SampleID = Unique sample identifier
Treatment = C (no OC, no prune), N (OC, no prune), P (OC, prune)
Day = 1 (day 1 of cycle), 21 (day 21 of cycle)
Age (years)
Ethnicity
Height (cm)
Weight (kg) = from DXA bone density scan (dual-energy x-ray absorption)
Calculated BMI 
AgeOfMenarche (years) = subject's age at the time of first menstruation
AvgNumDaysBtwPeriods = average number of days between periods
OCE2Dosage = oral contraceptive ethinyl estradiol dosage (mcg)
AgeFirstOCUse = age of first oral contraceptive use (years)
TotalYearsOnOC = total number of years on OC (years)
BowelDiaryCollectionTime = visit that data was collected; 1 = after Visit 1 (before Visit 2); 2 = after Visit 2 (Day 1 cycle or Day 21 of pill pack); 3 = after Visit 3 (Day 4 or Day 24 of pill pack); 4 =  after Visit 4 (Day 11 or Day 3 of pill pack); 5 = after Visit 5 (Day 21 or Day 13 of pill pack)
BMperDay = average number of bowel movements per day
BMperWk = number of bowel movements per week
AvgAmountPerWeek = average amount of stool produced in 7 days (cups)
AvgCon =  question #2; rating scale 1-7 stool consistency (1 = very soft; 7 = very hard) averaged for days they had bowel movements
AvgStrain = question #3; rating scale 1-7 straining during bowel movement (1 = none; 7 = extreme) averaged for days they had bowel movements
AvgPain= question 4; rating scale 1-7 pain during bowel movement(1 = none; 7 = extreme) averaged for days they had bowel movements 
AvgFCons= question #5; rating scale 1-7 overall feeling of constipation (1 = not constipated; 7 = very constipated) averaged for days they had bowel movements
NA = missing bowel diary

```{r} 
# How does oral contraceptive (OC) use affect the gut microbiome composition and diversity in women?
### Does OC use have an impact on the gut microbial diversity? (alpha and beta)?
### Expect to see differences in alpha diversity (differences in composition)

# Does OC use have an impact on the relative abundance of bacterial taxa in the gut microbiome?

# Are there any specific microbial taxa or diversity patterns associated with the use of OCs and/or prune supplementation?
### Lower relative abundance of Bifidobacteria, Lactobacillus, and Roseburia in women taking oral contraceptives

# Do prunes have an effect on the gut microbiome composition and diversity of women taking oral contraceptives?
### Expect to see a shift in the microbiome when adding prune
### Microbiome of OC users shifts back to that of non-users?
### Increase in Bifidobacteria, Lactobacillus, and Roseburia.


```
# 1. Import data
## 1.1 Import Metadata
```{r}

# #merge participant data, breath data, and barcode data to have a complete metadata file -- save and upload to drive
# participant_data <- read.csv("ParticipantData.csv", header = TRUE)
# head(participant_data)
# barcode_data <- read.csv("BarcodeData.tsv", header = TRUE, sep = "\t")
# head(barcode_data)
# map_all <- merge(participant_data, barcode_data, by = "SampleID", all.y = TRUE, all.x = TRUE)
# head(map_all)
# #
# #get column names
# colnames(map_all)
# #reorder columns to have barcodes in front
# map_all <- map_all[, c(1, 23, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22)]
# head(map_all)
# #
# breath_data <- read.csv("BreathData.tsv", header = TRUE, sep = "\t")
# head(breath_data)
# #
# map_all <- merge(map_all, breath_data, by = "SampleID", all.y = TRUE, all.x = TRUE)
# #
# #export df as .tsv file
# write_tsv(map_all, "metadata-r.tsv")
# # #take a look at NA values and change if needed


# import file created above
#map_all <- read.csv("metadata-r.tsv", header = TRUE, sep = "\t", row.names = 1)
# or import file with adjusted NAs
# map_all <- read.csv("metadata-r-na.tsv", header = TRUE, sep = "\t", row.names = 1)
map_all <- read.csv("metadata-r-na_treat-diet.tsv", header = TRUE, sep = "\t", row.names = 1) # N-Day1 combined with P-Day1

# Select columns needed from map
map_all <- map_all[c("SampleID","Subject","Treatment","Diet","Day","H2","CH4")]

#remove samples that were excluded due to denoising/quality filtering or low sequence depth, as well as controls:
# - "744P2", "983N2" removed due to low sequence frequencing
# - "920C2", "966N2", "916C2" removed during rarefaction
#if one:
# map <- map_all[!(map_all$SampleID=="1fe19636.3708.4a87.8e5a.cf10d8530d4e"),]
#if multiple:
remove_id <- c("negative2", "negative4", "Adrian", "ZymoE", "ZymoPCR", "920C2", "966N2", "916C2", "744P2","983N2")
map <- map_all[!(row.names(map_all) %in% remove_id),]

```
## 1.2 ASV table
```{r}
asv_16S_all <- read.delim("zero-filtered_data/ASV_table.zerofiltered.tsv", header = TRUE, row.names = 1, check.names = F)

head(asv_16S_all)

asv_16S <- asv_16S_all[,!(colnames(asv_16S_all) %in% remove_id)]
head(asv_16S)

```
## 1.3 Import Genus Table
```{r}
genus_all <- read.delim("zero-filtered_data/genus_table_from_biom.tsv", header = TRUE, row.names = 1, check.names = F)

genus <- genus_all[,!(colnames(genus_all) %in% remove_id)]
head(genus)

```
## 1.3 Import Species Table
```{r}
species_all <- read.delim("zero-filtered_data/species_table_from_biom.tsv", header = TRUE, row.names = 1, check.names = F)

species <- species_all[,!(colnames(species_all) %in% remove_id)]
head(species)

```
## 1.4 Import Alpha Diversity data
```{r}
# alpha_diversity_all <- read.csv("alpha_diversity-zerofiltered.csv", header = TRUE)
shannon <- read.delim("zero-filtered_data/shannon_entropy.tsv", header = TRUE)
observed_features <- read.delim("zero-filtered_data/observed_features.tsv", header = TRUE)
faith_pd <- read.delim("zero-filtered_data/faith_pd.tsv", header = TRUE)
pielou_evenness <- read.delim("zero-filtered_data/pielou_evenness.tsv", header = TRUE)

# Rename the sample ID column in each data frame so it matches on all data frames
colnames(shannon)[1] <- "SampleID"
colnames(observed_features)[1] <- "SampleID"
colnames(faith_pd)[1] <- "SampleID"
colnames(pielou_evenness)[1] <- "SampleID"

# Merge the data frames
alpha_diversity_all <- merge(shannon, observed_features, by = "SampleID")
alpha_diversity_all <- merge(alpha_diversity_all, faith_pd, by = "SampleID")
alpha_diversity_all <- merge(alpha_diversity_all, pielou_evenness, by = "SampleID")

# only for consistency:
alpha_diversity <- alpha_diversity_all

```
## 1.5 Import Unifrac Matrices
exported from qiime2, use unfiltered
```{r}
# weighted_unfiltered_all <- read.csv("zero-filtered_data/weighted_unifrac_distance_matrix.tsv", header = TRUE, row.names = 1, sep = "\t", check.names = F)
weighted_unfiltered_all <- read.csv("zero-filtered_data/weighted_unifrac_distance_matrix.tsv", header = TRUE, row.names = 1, sep = "\t", check.names = F)
head(weighted_unfiltered_all)

weighted_unfiltered <- weighted_unfiltered_all[!(rownames(weighted_unfiltered_all) %in% remove_id),]
weighted_unfiltered <- weighted_unfiltered[,!(colnames(weighted_unfiltered) %in% remove_id)]

```

```{r}
# unweighted_unfiltered_all <- read.csv("uzero-filtered_data/unweighted_unifrac_distance_matrix.tsv", header = TRUE, row.names = 1, sep = "\t", check.names = F)
unweighted_unfiltered_all <- read.csv("zero-filtered_data/unweighted_unifrac_distance_matrix.tsv", header = TRUE, row.names = 1, sep = "\t", check.names = F)
head(unweighted_unfiltered_all)

unweighted_unfiltered <- unweighted_unfiltered_all[!(rownames(unweighted_unfiltered_all) %in% remove_id),]
unweighted_unfiltered <- unweighted_unfiltered[,!(colnames(unweighted_unfiltered) %in% remove_id)]

```
check dimensions
```{r}
dim(weighted_unfiltered)
dim(unweighted_unfiltered)

```
## 1.6 Import taxonomy table into R and format to work with phyloseq
```{r}
# Read the taxonomy table (assuming it's in a CSV or TSV format):
taxonomy_raw <- read.table("taxonomy.tsv", stringsAsFactors = FALSE, header = TRUE, sep = "\t")
head(taxonomy_raw)

# Remove prefixes from Taxon names
taxonomy_raw$Taxon <- gsub("d__|p__|c__|o__|f__|g__|s__", "", taxonomy_raw$Taxon)
head(taxonomy_raw)

```
Split the taxonomic ranks and create a new taxonomy table:
```{r}
# # Convert the 'Taxon' column to a character before splitting (may not need to do this)
# taxonomy_raw$Taxon <- as.character(taxonomy_raw$Taxon)

# Split the "Taxon" column into taxonomic ranks
tax_ranks <- strsplit(taxonomy_raw$Taxon, "; ")

# Create a data frame with separate columns for each taxonomic rank
tax_table <- do.call(rbind, lapply(tax_ranks, function(x) {
  c(Domain = x[1], Phylum = x[2], Class = x[3], Order = x[4], Family = x[5], Genus = x[6], Species = x[7])
}))

# Set the row names of the tax_table to match the ASV/OTU names
rownames(tax_table) <- taxonomy_raw$Feature.ID
head(tax_table)

```
## 1.7 Import phylogenetic tree into R
```{r}
# Use read.tree from the 'ape' package to import Newick files
phylo_tree_rooted <- read.tree("phylogeny/rooted-tree.nwk")
phylo_tree_unrooted <- read.tree("phylogeny/unrooted-tree.nwk")

```
## 1.8 Subset metadata into Treatment and Cycle day groups
```{r}
# Check number of samples per group
table(map_all$Treatment, map_all$Day)
table(map_all$Diet, map_all$Day)

number_subjects <- length(unique(map_all$Subject))
number_subjects

# Create each group to use for alpha diversity and subsequent analyses

# Control vs. OC Users
group_control_oc <- subset(map, Treatment %in% c("Control", "User") & Diet %in% c("No Prune"))
# Control
group_control <- subset(map, Treatment %in% c("Control") & Diet %in% c("No Prune"))
# OC User
group_user <- subset(map, Treatment %in% c("User"))

# Prune vs. No Prune
group_prune_no.prune <- subset(map, Treatment %in% c("User") & Diet %in% c("Prune", "No Prune"))
# Prune
group_prune <- subset(map, Treatment %in% c("User") & Diet %in% c("Prune"))
# No Prune
group_no.prune <- subset(map, Treatment %in% c("User") & Diet %in% c("No Prune"))

# group_prune_day1.21 <- group_prune_no.prune %>%
#   group_by(Subject) %>%
#   filter(any(Day == "Day 1") & any(Day == "Day 21")) %>%
#   ungroup

# # Subset the data
# group_prune_day1.21 <- group_prune_no.prune %>%
#   group_by(Subject) %>%
#   filter(
#     any(Diet == "No Prune" & Day == "Day 1") & 
#     any(Diet == "Prune" & Day == "Day 21")
#   ) %>%
#   ungroup()

```
